lib/web/websocket/receiver.js

```diff
@@ -298,26 +298,28 @@ class ByteParser extends Writable {
       // defined as the status code (Section 7.4) contained in the first Close
       // control frame received by the application
       code = data.readUInt16BE(0)
+      if (!isValidStatusCode(code)) {
+        return null
+      }
+    } else {
+      return null
     }
 
     // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.6
-    /** @type {Buffer} */
-    let reason = data.subarray(2)
+    /** @type {number} */
+    let reasonOffset = 2
 
-    // Remove BOM
-    if (reason[0] === 0xEF && reason[1] === 0xBB && reason[2] === 0xBF) {
-      reason = reason.subarray(3)
+    // Detect BOM and ignore it
+    if (data.length >= 5 && data[2] === 0xEF && data[3] === 0xBB && data[4] === 0xBF) {
+      reasonOffset = 5
     }
 
-    if (code !== undefined && !isValidStatusCode(code)) {
-      return null
+    if (data.length - reasonOffset < 1) {
+      return { code, reason: '' }
     }
 
-    try {
-      reason = utf8Decode(reason)
-    } catch {
-      return null
-    }
+    /** @type {string} */
+    const reason = utf8DecodeSafe(data.subarray(reasonOffset))
```

Comment:
This is wrong. If we receive an invalid utf-8 reason we are meant to fail the websocket connection.

Reply:
The reason we don't fail it in that spot is because it's already closing, but the logic is completely wrong here, and will lead to issues if we do check the return type.